a) In be67c41168 -> this will not be bisectable since it will fail to build.:

```
  CC       staticd/static_nht.o
pimd/pim_tib.c: In function ‘tib_sg_proxy_join_prune_check’:
pimd/pim_tib.c:101:30: error: too few arguments to function ‘pim_filter_match’
  101 |                         if (!pim_filter_match(&pim_ifp->gm_proxy_filter, &pfx, ifp))
      |                              ^~~~~~~~~~~~~~~~
In file included from pimd/pim_instance.h:18,
                 from pimd/pim_tib.c:12:
pimd/pim_routemap.h:32:13: note: declared here
   32 | extern bool pim_filter_match(const struct pim_filter_ref *ref, const struct prefix_sg *sg,
      |             ^~~~~~~~~~~~~~~~
  CC       staticd/static_routes.o
  CC       staticd/static_zebra.o
  CC       staticd/static_vrf.o
make[1]: *** [Makefile:13988: pimd/pimd-pim_tib.o] Error 1
make[1]: *** Waiting for unfinished jobs....
  CC       staticd/static_nb.o
pimd/pim_iface.c: In function ‘pim_if_gm_proxy_init’:
pimd/pim_iface.c:1624:38: error: too few arguments to function ‘pim_filter_match’
 1624 |                                 if (!pim_filter_match(&oif_pim->gm_proxy_filter, &pfx, oif))
      |                                      ^~~~~~~~~~~~~~~~
In file included from pimd/pim_instance.h:18,
                 from pimd/pim_iface.c:22:
pimd/pim_routemap.h:32:13: note: declared here
   32 | extern bool pim_filter_match(const struct pim_filter_ref *ref, const struct prefix_sg *sg,
      |             ^~~~~~~~~~~~~~~~
pimd/pim_iface.c: In function ‘pim_if_gm_proxy_init’:
pimd/pim_iface.c:1624:38: error: too few arguments to function ‘pim_filter_match’
 1624 |                                 if (!pim_filter_match(&oif_pim->gm_proxy_filter, &pfx, oif))
      |                                      ^~~~~~~~~~~~~~~~
In file included from pimd/pim_instance.h:18,
                 from pimd/pim_iface.c:22:
pimd/pim_routemap.h:32:13: note: declared here
   32 | extern bool pim_filter_match(const struct pim_filter_ref *ref, const struct prefix_sg *sg,
      |             ^~~~~~~~~~~~~~~~
make[1]: *** [Makefile:12878: pimd/pim6d-pim_iface.o] Error 1
pimd/pim_tib.c: In function ‘tib_sg_proxy_join_prune_check’:
pimd/pim_tib.c:101:30: error: too few arguments to function ‘pim_filter_match’
  101 |                         if (!pim_filter_match(&pim_ifp->gm_proxy_filter, &pfx, ifp))
      |                              ^~~~~~~~~~~~~~~~
In file included from pimd/pim_instance.h:18,
                 from pimd/pim_tib.c:12:
pimd/pim_routemap.h:32:13: note: declared here
   32 | extern bool pim_filter_match(const struct pim_filter_ref *ref, const struct prefix_sg *sg,
      |             ^~~~~~~~~~~~~~~~
make[1]: *** [Makefile:13630: pimd/pimd-pim_iface.o] Error 1
make[1]: *** [Makefile:13236: pimd/pim6d-pim_tib.o] Error 1
```

--- a/pimd/pim_cmd.c
+++ b/pimd/pim_cmd.c
@@ -6002,6 +6002,22 @@ DEFPY (interface_ip_igmp_proxy,
 	return pim_process_ip_gmp_proxy_cmd(vty, !no);
 }
 
+DEFPY_YANG(interface_ip_igmp_proxy_rmap, interface_ip_igmp_proxy_rmap_cmd,
+	   "[no] ip igmp proxy route-map ![RMAP_NAME]",
+	   NO_STR
+	   IP_STR
+	   IGMP_STR
+	   "Proxy IGMP join/prune operations\n"
+	   "Filter proxied groups through route-map\n"
+	   "Route-map name\n")
+{
+	if (no)
+		nb_cli_enqueue_change(vty, "./proxy-route-map", NB_OP_DESTROY, NULL);
+	else
+		nb_cli_enqueue_change(vty, "./proxy-route-map", NB_OP_MODIFY, rmap_name);
+
+	return nb_cli_apply_changes(vty, FRR_GMP_INTERFACE_XPATH, FRR_PIM_AF_XPATH_VAL);
+}
 
 DEFPY_YANG(interface_ip_pim_neighbor_prefix_list,
            interface_ip_pim_neighbor_prefix_list_cmd,
@@ -9410,6 +9426,7 @@ void pim_cmd_init(void)
 	install_element(INTERFACE_NODE,
 			&interface_no_ip_igmp_last_member_query_interval_cmd);
 	install_element(INTERFACE_NODE, &interface_ip_igmp_proxy_cmd);
+	install_element(INTERFACE_NODE, &interface_ip_igmp_proxy_rmap_cmd);
 	install_element(INTERFACE_NODE, &interface_ip_igmp_limits_cmd);
 	install_element(INTERFACE_NODE, &no_interface_ip_igmp_limits_cmd);
 	install_element(INTERFACE_NODE, &interface_ip_igmp_immediate_leave_cmd);

--- a/pimd/pim_iface.c
+++ b/pimd/pim_iface.c
@@ -39,6 +39,7 @@
 #include "pim_vxlan.h"
 #include "pim_tib.h"
 #include "pim_util.h"
+#include "pim_routemap.h"
 
 #include "pim6_mld.h"
 
@@ -187,6 +188,7 @@ struct pim_interface *pim_if_new(struct interface *ifp, bool gm, bool pim,
 	pim_ifp->pim->mcast_if_count++;
 
 	pim_filter_ref_init(&pim_ifp->gmp_filter);
+	pim_filter_ref_init(&pim_ifp->gm_proxy_filter);
 
 	return pim_ifp;
 }
@@ -227,6 +229,7 @@ void pim_if_delete(struct interface *ifp)
 	pim_igmp_if_fini(pim_ifp);
 
 	pim_filter_ref_fini(&pim_ifp->gmp_filter);
+	pim_filter_ref_fini(&pim_ifp->gm_proxy_filter);
 
 	list_delete(&pim_ifp->pim_neighbor_list);
 	list_delete(&pim_ifp->upstream_switch_list);
@@ -1595,6 +1598,7 @@ static void pim_if_static_group_del_all(struct interface *ifp)
 void pim_if_gm_proxy_init(struct pim_instance *pim, struct interface *oif)
 {
 	struct interface *ifp;
+	struct pim_interface *oif_pim = oif->info;
 
 	FOR_ALL_INTERFACES (pim->vrf, ifp) {
 		struct pim_interface *pim_ifp = ifp->info;
@@ -1612,6 +1616,18 @@ void pim_if_gm_proxy_init(struct pim_instance *pim, struct interface *oif)
 					  group)) {
 			for (ALL_LIST_ELEMENTS_RO(group->group_source_list,
 						  source_node, src)) {
+				pim_sgaddr sgaddr = { .src = src->source_addr,
+						      .grp = group->group_addr };
+				struct prefix_sg pfx;
+
+				pim_sg_to_prefix(&sgaddr, &pfx);
+				if (!pim_filter_match(&oif_pim->gm_proxy_filter, &pfx, oif)) {
+					if (PIM_DEBUG_GM_TRACE)
+						zlog_debug("%s: proxy join for SG%pPSG from %s to %s filtered due to route-map",
+							   __func__, &pfx, ifp->name, oif->name);
+					continue;
+				}
+
 				pim_if_gm_join_add(oif, group->group_addr,
 						   src->source_addr,
 						   GM_JOIN_PROXY);

--- a/pimd/pim_iface.h
+++ b/pimd/pim_iface.h
@@ -76,6 +76,7 @@ struct pim_interface {
 
 	bool gm_enable : 1;
 	bool gm_proxy : 1; /* proxy IGMP joins/prunes */
+	struct pim_filter_ref gm_proxy_filter; /* route-map filter for proxied (S,G) */
 
 	ifindex_t mroute_vif_index;
 	struct pim_instance *pim;

--- a/pimd/pim_nb.c
+++ b/pimd/pim_nb.c
@@ -885,6 +885,13 @@ const struct frr_yang_module_info frr_gmp_info = {
 			}
 		},
 		{
+			.xpath = "/frr-interface:lib/interface/frr-gmp:gmp/address-family/proxy-route-map",
+			.cbs = {
+				.modify  = lib_interface_gm_proxy_rmap_modify,
+				.destroy = lib_interface_gm_proxy_rmap_destroy,
+			}
+		},
+		{
 			.xpath = "/frr-interface:lib/interface/frr-gmp:gmp/address-family/immediate-leave",
 			.cbs = {
 				.modify = lib_interface_gmp_immediate_leave_modify,

--- a/pimd/pim_nb.h
+++ b/pimd/pim_nb.h
@@ -323,6 +323,8 @@ int lib_interface_gmp_immediate_leave_modify(struct nb_cb_modify_args *args);
 int lib_interface_gmp_require_router_alert_modify(struct nb_cb_modify_args *args);
 int lib_interface_gm_rmap_modify(struct nb_cb_modify_args *args);
 int lib_interface_gm_rmap_destroy(struct nb_cb_destroy_args *args);
+int lib_interface_gm_proxy_rmap_modify(struct nb_cb_modify_args *args);
+int lib_interface_gm_proxy_rmap_destroy(struct nb_cb_destroy_args *args);
 int lib_interface_gm_alist_modify(struct nb_cb_modify_args *args);
 int lib_interface_gm_alist_destroy(struct nb_cb_destroy_args *args);
 int lib_interface_pim_address_family_allow_rp_modify(struct nb_cb_modify_args *args);

--- a/pimd/pim_nb_config.c
+++ b/pimd/pim_nb_config.c
@@ -5093,6 +5093,59 @@ int lib_interface_gm_rmap_destroy(struct nb_cb_destroy_args *args)
 }
 
 /*
+ * XPath: /frr-interface:lib/interface/frr-gmp:gmp/address-family/proxy-route-map
+ */
+int lib_interface_gm_proxy_rmap_modify(struct nb_cb_modify_args *args)
+{
+	struct interface *ifp;
+	struct pim_interface *pim_ifp;
+	const char *rmap;
+
+	switch (args->event) {
+	case NB_EV_VALIDATE:
+	case NB_EV_ABORT:
+	case NB_EV_PREPARE:
+		break;
+	case NB_EV_APPLY:
+		ifp = nb_running_get_entry(args->dnode, NULL, true);
+		pim_ifp = ifp->info;
+		if (!pim_ifp) {
+			pim_ifp = pim_if_new(ifp, true, false, false, false);
+			ifp->info = pim_ifp;
+		}
+
+		rmap = yang_dnode_get_string(args->dnode, NULL);
+		pim_filter_ref_set_rmap(&pim_ifp->gm_proxy_filter, rmap);
+		break;
+	}
+
+	return NB_OK;
+}
+
+int lib_interface_gm_proxy_rmap_destroy(struct nb_cb_destroy_args *args)
+{
+	struct interface *ifp;
+	struct pim_interface *pim_ifp;
+
+	switch (args->event) {
+	case NB_EV_VALIDATE:
+	case NB_EV_ABORT:
+	case NB_EV_PREPARE:
+		break;
+	case NB_EV_APPLY:
+		ifp = nb_running_get_entry(args->dnode, NULL, true);
+		pim_ifp = ifp->info;
+		if (!pim_ifp)
+			return NB_ERR_INCONSISTENCY;
+
+		pim_filter_ref_set_rmap(&pim_ifp->gm_proxy_filter, NULL);
+		break;
+	}
+
+	return NB_OK;
+}
+
+/*
  * XPath: /frr-interface:lib/interface/frr-gmp:gmp/address-family/access-list
  */
 int lib_interface_gm_alist_modify(struct nb_cb_modify_args *args)

--- a/pimd/pim_tib.c
+++ b/pimd/pim_tib.c
@@ -15,6 +15,7 @@
 #include "pim_oil.h"
 #include "pim_nht.h"
 #include "pim_dm.h"
+#include "pim_routemap.h"
 
 static struct channel_oil *
 tib_sg_oil_setup(struct pim_instance *pim, pim_sgaddr sg, struct interface *oif)
@@ -94,6 +95,20 @@ void tib_sg_proxy_join_prune_check(struct pim_instance *pim, pim_sgaddr sg,
 			continue;
 
 		if (pim_ifp->gm_enable && pim_ifp->gm_proxy) {
+			struct prefix_sg pfx;
+
+			pim_sg_to_prefix(&sg, &pfx);
+			/* Apply the proxy route-map only to joins.  Prunes must always run
+			 * so tightening the filter without cycling the proxy cannot strand
+			 * proxy (*,G)/(S,G) state after the last host leaves.
+			 */
+			if (join && !pim_filter_match(&pim_ifp->gm_proxy_filter, &pfx, ifp)) {
+				if (PIM_DEBUG_GM_TRACE)
+					zlog_debug("%s: proxy join for SG%pPSG from %s to %s filtered due to route-map",
+						   __func__, &pfx, oif->name, ifp->name);
+				continue;
+			}
+
 			if (join)
 				pim_if_gm_join_add(ifp, sg.grp, sg.src,
 						   GM_JOIN_PROXY);

--- a/pimd/pim_vty.c
+++ b/pimd/pim_vty.c
@@ -307,11 +307,6 @@ static int gm_config_write(struct vty *vty, int writes,
 		++writes;
 	}
 
-	if (pim_ifp->gm_proxy) {
-		vty_out(vty, " ip igmp proxy\n");
-		++writes;
-	}
-
 	/* ip igmp version */
 	if (pim_ifp->igmp_version != IGMP_DEFAULT_VERSION) {
 		vty_out(vty, " ip igmp version %d\n", pim_ifp->igmp_version);
@@ -507,6 +502,24 @@ int pim_config_write(struct vty *vty, int writes, struct interface *ifp,
 	if (pim_ifp->gmp_filter.alistname) {
 		vty_out(vty, " " PIM_AF_NAME " " GM_AF_DBG " access-list %s\n",
 			pim_ifp->gmp_filter.alistname);
+		++writes;
+	}
+
+	/*
+	 * IF igmp/mld proxy route-map
+	 *
+	 * Must be emitted before 'proxy' so that on config replay
+	 * pim_if_gm_proxy_init() runs with the filter already set.
+	 */
+	if (pim_ifp->gm_proxy_filter.rmapname) {
+		vty_out(vty, " " PIM_AF_NAME " " GM_AF_DBG " proxy route-map %s\n",
+			pim_ifp->gm_proxy_filter.rmapname);
+		++writes;
+	}
+
+	/* IF igmp/mld proxy */
+	if (pim_ifp->gm_proxy) {
+		vty_out(vty, " " PIM_AF_NAME " " GM_AF_DBG " proxy\n");
 		++writes;
 	}
 

--- a/yang/frr-gmp.yang
+++ b/yang/frr-gmp.yang
@@ -171,6 +171,12 @@ module frr-gmp {
         "Enable IGMP proxy on the interface.";
     }
 
+    leaf proxy-route-map {
+      type frr-route-map:route-map-ref;
+      description
+        "Filter proxied multicast (S,G) groups with referenced route-map.";
+    }
+
     leaf max-groups {
       type uint32;
       default "4294967295";

